Mathlib/Analysis/Distribution/DerivNotation.lean

```diff
@@ -0,0 +1,184 @@
+/-
+Copyright (c) 2025 Moritz Doll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Doll
+-/
+module
+
+public import Mathlib.Algebra.Module.Equiv.Defs
+public import Mathlib.Data.Fin.Tuple.Basic
+public import Mathlib.Topology.Algebra.Module.LinearMap
+
+/-! # Type classes for derivatives
+
+In this file we define notation type classes for line derivatives, also known as partial
+derivatives.
+
+Moreover, we provide type-classes that encode the linear structure.
+
+Currently, this type class is only used by Schwartz functions. Future uses include derivatives on
+test functions, distributions, tempered distributions, and Sobolev spaces (and other generalized
+function spaces).
+-/
+
+@[expose] public section
+
+universe u' u v w
+
+variable {R V E F : Type*}
+
+open Fin
+
+/--
+The notation typeclass for the line derivative.
+-/
+class LineDeriv (V : Type u) (E : Type v) (F : outParam (Type w)) where
+  /-- `∂_{v} f` is the line derivative of `f` in direction `v`. The meaning of this notation is
+  type-dependent. -/
+  lineDerivOp : V → E → F
+
+namespace LineDeriv
+
+@[inherit_doc] scoped notation "∂_{" v "}" => LineDeriv.lineDerivOp v
+
+variable {V E : Type*} [LineDeriv V E E]
+
+/-- `∂^{m} f` is the iterated line derivative of `f`. -/
+def iteratedLineDerivOp {n : ℕ} : (Fin n → V) → E → E :=
+  Nat.recOn n (fun _ ↦ id) (fun _ rec y ↦ LineDeriv.lineDerivOp (y 0) ∘ rec (tail y))
+
+@[inherit_doc] scoped notation "∂^{" v "}" => LineDeriv.iteratedLineDerivOp v
+
+@[simp]
+theorem iteratedLineDerivOp_zero (m : Fin 0 → V) (f : E) : ∂^{m} f = f :=
+  rfl
+
+@[simp]
+theorem iteratedLineDerivOp_one (m : Fin 1 → V) (f : E) : ∂^{m} f = ∂_{m 0} f :=
+  rfl
+
+theorem iteratedLineDerivOp_succ_left {n : ℕ} (m : Fin (n + 1) → V) (f : E) :
+    ∂^{m} f = ∂_{m 0} (∂^{tail m} f) :=
+  rfl
+
+theorem iteratedLineDerivOp_succ_right {n : ℕ} (m : Fin (n + 1) → V) (f : E) :
+    ∂^{m} f = ∂^{init m} (∂_{m (last n)} f) := by
+  induction n with
+  | zero => rfl
+  -- The proof is `∂^{n + 2} = ∂ ∂^{n + 1} = ∂ ∂^n ∂ = ∂^{n+1} ∂`
+  | succ n IH =>
+    have hmzero : init m 0 = m 0 := by simp only [init_def, castSucc_zero]
+    have hmtail : tail m (last n) = m (last n.succ) := by
+      simp only [tail_def, succ_last]
+    calc
+      _ = ∂_{m 0} (∂^{tail m} f) := iteratedLineDerivOp_succ_left _ _
+      _ = ∂_{m 0} (∂^{init <| tail m} (∂_{tail m <| last n} f)) := by
+        congr 1
+        exact IH _
+      _ = _ := by
+        rw [hmtail, iteratedLineDerivOp_succ_left, hmzero, tail_init_eq_init_tail]
+
+end LineDeriv
+
+open LineDeriv
+
+/--
+The line derivative is additive, `∂_{v} (x + y) = ∂_{v} x + ∂_{v} y` for all `x y : E`.
```

Comment:
Please mention that for the usual `lineDeriv`, it holds for the derivatives of bundled smooth functions only.